ui/apps/dashboard/src/app/(dashboard)/env/[environmentSlug]/functions/useRows.ts

```diff
@@ -0,0 +1,94 @@
+import { useEffect, useState } from 'react';
+import { useClient } from 'urql';
+
+import { useEnvironment } from '@/app/(dashboard)/env/[environmentSlug]/environment-context';
+import { getFunctionUsagesPage, useFunctionsPage } from '@/queries';
+import type { FunctionTableRow } from './FunctionTable';
+
+type Output = {
+  error: Error | undefined;
+  hasMore: boolean | undefined;
+  isLoading: boolean;
+  loadMore: () => void;
+  rows: FunctionTableRow[];
+};
+
+export function useRows({ archived }: { archived: boolean }): Output {
+  const client = useClient();
+  const env = useEnvironment();
+  const [page, setPage] = useState(1);
+
+  // Primary function data. This doesn't include usage data since we'll
+  // lazy-load it.
+  const [functionsData, setFunctionsData] = useState<{
+    lastPage: number;
+    rows: FunctionTableRow[];
+  }>({
+    lastPage: 0,
+    rows: [],
+  });
+
+  // Reset function data when switching between archived and active.
+  useEffect(() => {
+    setFunctionsData({
+      lastPage: 0,
+      rows: [],
+    });
+  }, [archived]);
+
+  const functionsRes = useFunctionsPage({
+    archived,
+    envID: env.id,
+    page,
+  });
+
+  // Append new function data.
+  useEffect(() => {
+    if (functionsRes.data && functionsRes.data.page.page > functionsData.lastPage) {
```

Comment:
Does this have to be `functionsRes.data.page.page` ? With the same name it's difficult to understand what each layer returns

Reply:
That's what our GraphQL endpoint returns. Maybe we can fix that when we standardize pagination?